File: objectivec/day20_part2_2022.m
#import <Foundation/Foundation.h>

typedef struct {
    NSInteger pos, val;
} num;

void mix(num *nums, NSInteger count);
NSInteger coords(num *nums, NSInteger count);
NSString* readAll(NSString *path);
NSInteger toInt(NSString *s);

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = readAll(@"input.txt");
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        NSInteger count = lines.count;
        num nums[count];
        
        for (NSInteger i = 0; i < count; i++) {
            nums[i].pos = i;
            nums[i].val = toInt(lines[i]);
        }

        num nums2[count];
        for (NSInteger i = 0; i < count; i++) {
            nums2[i] = (num){nums[i].pos, 811589153 * nums[i].val};
        }

        for (NSInteger i = 0; i < 10; i++) {
            mix(nums2, count);
        }
        
        NSLog(@"%ld", (long)coords(nums2, count));
    }
    return 0;
}

void mix(num *nums, NSInteger count) {
    NSInteger n = count - 1;
    for (NSInteger i = 0; i < count; i++) {
        NSInteger oldpos = nums[i].pos;
        NSInteger newpos = ((oldpos + nums[i].val) % n + n) % n;
        if (oldpos < newpos) {
            for (NSInteger j = 0; j < count; j++) {
                if (nums[j].pos > oldpos && nums[j].pos <= newpos) {
                    nums[j].pos--;
                }
            }
        }
        if (newpos < oldpos) {
            for (NSInteger j = 0; j < count; j++) {
                if (nums[j].pos >= newpos && nums[j].pos < oldpos) {
                    nums[j].pos++;
                }
            }
        }
        nums[i].pos = newpos;
    }
}

NSInteger coords(num *nums, NSInteger count) {
    NSInteger zeroPos = 0;
    for (NSInteger i = 0; i < count; i++) {
        if (nums[i].val == 0) {
            zeroPos = nums[i].pos;
            break;
        }
    }
    NSInteger sum = 0;
    for (NSInteger i = 0; i < count; i++) {
        if (nums[i].pos == (zeroPos + 1000) % count || 
            nums[i].pos == (zeroPos + 2000) % count || 
            nums[i].pos == (zeroPos + 3000) % count) {
            sum += nums[i].val;
        }
    }
    return sum;
}

NSString* readAll(NSString *path) {
    NSError *error = nil;
    NSString *content = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&error];
    if (error) {
        @throw [NSException exceptionWithName:@"FileReadException" reason:error.localizedDescription userInfo:nil];
    }
    return [content stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
}

NSInteger toInt(NSString *s) {
    return [s integerValue];
}

File: c/day20_part2_2022.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    int pos;
    long long val;
} num;

void mix(num *nums, int n) {
    for (int i = 0; i < n; i++) {
        int oldpos = nums[i].pos;
        long long newpos = ((long long)oldpos + nums[i].val) % (n - 1);
        if (newpos < 0) newpos += (n - 1);
        
        if (oldpos < newpos) {
            for (int j = 0; j < n; j++) {
                if (nums[j].pos > oldpos && nums[j].pos <= newpos) {
                    nums[j].pos--;
                }
            }
        } else if (newpos < oldpos) {
            for (int j = 0; j < n; j++) {
                if (nums[j].pos >= newpos && nums[j].pos < oldpos) {
                    nums[j].pos++;
                }
            }
        }
        nums[i].pos = newpos;
    }
}

long long coords(num *nums, int n) {
    int zeroPos = 0;
    for (int i = 0; i < n; i++) {
        if (nums[i].val == 0) {
            zeroPos = nums[i].pos;
            break;
        }
    }
    long long sum = 0;
    for (int i = 0; i < n; i++) {
        if (nums[i].pos == (zeroPos + 1000) % n || 
            nums[i].pos == (zeroPos + 2000) % n ||
            nums[i].pos == (zeroPos + 3000) % n) {
            sum += nums[i].val;
        }
    }
    return sum;
}

int main() {
    FILE *fp = fopen("input.txt", "r");
    if (fp == NULL) {
        perror("Error opening file");
        return 1;
    }

    char *line = NULL;
    size_t len = 0;
    ssize_t read;
    int count = 0;
    num *nums = NULL;

    while ((read = getline(&line, &len, fp)) != -1) {
        nums = realloc(nums, (count + 1) * sizeof(num));
        if (nums == NULL) {
            perror("Realloc failed");
            fclose(fp);
            if(line) free(line);
            return 1;
        }
        
        nums[count].pos = count;
        nums[count].val = atoll(line) * 811589153;
        count++;
    }

    fclose(fp);
    if(line) free(line);

    for (int i = 0; i < 10; i++) {
        mix(nums, count);
    }
    
    printf("%lld\n", coords(nums, count));
    free(nums);
    return 0;
}
